Update Feature/Improve gathering with more parameters

ebork11 opened the pull request:
Added ability to require an item either in the players hands or just in their inventory in order to gather. Also added the ability to gather multiple from one interaction instead of always just one.

LuxPoseidon reviewed (commented):
It may be better to assign your local variables as member variables in the class. (see what others think)
Also on line 67 you create and assign a local var playerController but it is never used....

Other than that I don't see anything inherently wrong.

ebork11 commented:
> It may be better to assign your local variables as member variables in the class. (see what others think) Also on line 67 you create and assign a local var playerController but it is never used....
> 
> Other than that I don't see anything inherently wrong.

Went ahead and removed that unnecessary line and changed the variable to match the member variables case. Thanks!

ebork11 commented:
![image](https://user-images.githubusercontent.com/43502340/172295243-4a7147e1-0e09-4fbb-b5a3-be16087fe65d.png)
![image](https://user-images.githubusercontent.com/43502340/172295263-a3f371b4-89c6-4610-ae7a-455042bf0730.png)
- Changed messages when unavailable so they are more informative.
- Also changed default value to 1 for amount given to prevent confusion on it being 0.

Arkensor commented on src/Scripts/Game/Features/Gathering/EL_GatherAction.c:
```diff
@@ -5,31 +5,114 @@ class EL_GatherAction : ScriptedUserAction
+	[Attribute(defvalue:"1", UIWidgets.EditBox, desc: "Amount of items to receive")]
+	private int m_AmountGathered;
+	
+	[Attribute("", UIWidgets.EditBox, desc: "Display name of what is needed")]
+	private string m_RequiredItemDisplayName;
```
This can/should be retrieved from the prefab, so it uses a localized display name of whatever item is being linked. We will need to figure out a good system to cache display names etc for different assets, as to get the info I think we might need to spawn them locally at least once.

ebork11 replied:
I can grab the prefab name from inside the script and then format it. As long as all prefabs are named correctly that should work fine.

Arkensor replied:
Ah no, was talking about the UIInfo on the inventoryitemcomponent of each of the items. Access to that is easy but would be wasteful to set up every time the action checks it. I will have to do some testing for this before I can give you any reliabe suggestion

ebork11 replied:
In the mean time should I remove the Attribute and have it just say "Item Required" until this is figured out. That way other people can start using the updated Gather function?

Arkensor replied:
Yup, that is a good idea. I am working on a prefab -> display name cache in the mean time and will add it as soon as its ready :)

ebork11 replied:
@Arkensor just updated it to remove that attribute and just default say "Requires item" or "Requires item in hand" for now until the cache is implemented!

Arkensor commented on src/Scripts/Game/Features/Gathering/EL_GatherAction.c:
```diff
@@ -5,31 +5,114 @@ class EL_GatherAction : ScriptedUserAction
+	[Attribute("", UIWidgets.CheckBox, desc: "Check entire inventory too")]
+	private bool m_CheckInventory;
 		
+	[Attribute(defvalue:"0", UIWidgets.EditBox, desc: "Amount of delay after performing action")]
+	private int m_DelayTimeMilliseconds;
```
What is the idea behind the delay? If it is to limit the gathering speed, then that is instead controlled by how long the action takes isn't it?

ebork11 replied:
One use I thought of for a delay would be something like robbing a trader. You wouldn't want a user to be able to spam the gather function which would require a separate timer from the duration one.

Another possibility I was thinking of adding was the ability to toggle the entities visibility while it's on delay so that it is hidden, but reappears.

Arkensor commented on src/Scripts/Game/Features/Gathering/EL_GatherAction.c:
```diff
@@ -5,31 +5,114 @@ class EL_GatherAction : ScriptedUserAction
+		{
+			SetCannotPerformReason(string.Format("Requires %1", m_RequiredItemDisplayName));
+			
+			array<BaseInventoryStorageComponent> outStorages = new array<BaseInventoryStorageComponent>();
+			m_InventoryManager.GetStorages(outStorages);
+			
+			foreach (BaseInventoryStorageComponent storage : outStorages)
+			{
+				array<IEntity> outItems = new array<IEntity>();
+				storage.GetAll(outItems);
+				
+				foreach (IEntity item : outItems)
+				{
+					if (item.GetPrefabData().GetPrefabName() == m_RequiredItemPrefab)
+						return true;
+				}
+			}
```
You can use `m_InventoryManager.GetItems()` instead of loading the storages and then the items manually :)
